Run some pde.build tests with Tycho (second attempt)

laeubi opened the pull request:
This is a fresh attempt based on

- https://github.com/eclipse-pde/eclipse.pde/pull/927

with some progress made here
- https://github.com/eclipse-pde/eclipse.pde/pull/933

but that was a dead end I think instead one should try to add a `category.xml` with another execution that mirrors into the "deltapack folder" so if the test executes in `target/work` the goal would be to have a folder `target/deltapack/eclipse` that mirrors everything needed as the "deltapack".

iloveeclipse commented:
`org.eclipse.pde.build.tests: Unqualified OSGi version 1.4.300.qualifier must match unqualified Maven version 1.4.200-SNAPSHOT`

laeubi commented:
> `org.eclipse.pde.build.tests: Unqualified OSGi version 1.4.300.qualifier must match unqualified Maven version 1.4.200-SNAPSHOT`

Seems it was too late yesterday, I fixed that now and also added an assumption to the launcher find method, this seems for me the most suitable solution because then we can merge this attempt already and add more work (e.g. enable more test) on top of it.

laeubi commented:
This seems to reproduce the test failures we see here
- https://github.com/eclipse-pde/eclipse.pde/issues/967

I therefore think this PR should be merged so we can see them in regular verification builds and can fix/act on them... in addition to that it adds 35 new test we will be aware if they would break.

@HannesWell @vik-chand any concerns?

HannesWell commented on build/org.eclipse.pde.build.tests/pom.xml:
```diff
@@ -0,0 +1,57 @@
+				<configuration>
+					<product>org.eclipse.sdk.ide</product>
+					<testRuntime>p2Installed</testRuntime>
+					<testClass>org.eclipse.pde.build.tests.PDEBuildTestSuite</testClass>
```
Am I right, that this testClass is used for the default `test` goal execution of the `tycho-surefire-plugin` for `eclipse-test-plugins` and the overwritten value in the separate `p2-tests` execution refelects the second execution defined in this projects test.xml?

laeubi replied:
Yes

HannesWell commented on build/org.eclipse.pde.build.tests/pom.xml:
```diff
@@ -0,0 +1,57 @@
+						<goals><goal>test</goal></goals>
+						<configuration>
+							<testClass>org.eclipse.pde.build.tests.P2TestSuite</testClass>
+							<systemProperties/>
```
```suggestion
```
Is this necessary? The effective pom view says this element is not configured in a parent pom.

laeubi replied:
This is to clear the inherited properties from the default execution/plugin configuration, one might can configure the default execution but lets dont complicate it too much as it already is :-)

HannesWell replied:
> but lets dont complicate it too much as it already is :-)

That's fine :)
But depending on where the to clear property is defined and how this might not be sufficient and something like https://stackoverflow.com/a/33252763/14542697 needs to be applied.
I have not further checked where this is coming from exactly, but I'm sure you are aware of all that.
Just wanted to make sure, it is not just forgotten.

HannesWell reviewed (approved):
> This seems to reproduce the test failures we see here
> 
>     * [120 P2Tests fail on master since I20231201-1800 #967](https://github.com/eclipse-pde/eclipse.pde/issues/967)
> 

Yes, I checked a few failures and they are identical.

> 
> I therefore think this PR should be merged so we can see them in regular verification builds and can fix/act on them... in addition to that it adds 35 new test we will be aware if they would break.
> 

It's not ideal but sounds like a suitable solution for the current situation. It will also help to verify a fix sooner.

Thank you @laeubi for making this possible so that we can detect regressions early.
Hopefully this prevent future regressions and will reduce the burden for @iloveeclipse to hunt down their causes.

iloveeclipse commented:
Thanks for enabling tests! Was a long outstanding PDE issue.

akurtakov commented:
I just saw aggregator build failing https://ci.eclipse.org/platform/job/eclipse.platform.releng.aggregator/job/PR-1639/1/console :
```
22:23:57  org.apache.maven.ProjectCycleException: The projects in the reactor contain a cyclic reference: Edge between 'Vertex{label='eclipse.platform.repository:eclipse.platform.repository:4.31.0-SNAPSHOT'}' and 'Vertex{label='org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT'}' introduces to cycle in the graph org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT --> org.eclipse.pde:org.eclipse.pde.build.tests:1.4.300-SNAPSHOT --> eclipse.platform.repository:eclipse.platform.repository:4.31.0-SNAPSHOT --> org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT
```
I'm not sure it's caused by this one here but looks suspiciously connected.

iloveeclipse commented:
> I just saw aggregator build failing https://ci.eclipse.org/platform/job/eclipse.platform.releng.aggregator/job/PR-1639/1/console :
> 
> ```
> 22:23:57  org.apache.maven.ProjectCycleException: The projects in the reactor contain a cyclic reference: Edge between 'Vertex{label='eclipse.platform.repository:eclipse.platform.repository:4.31.0-SNAPSHOT'}' and 'Vertex{label='org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT'}' introduces to cycle in the graph org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT --> org.eclipse.pde:org.eclipse.pde.build.tests:1.4.300-SNAPSHOT --> eclipse.platform.repository:eclipse.platform.repository:4.31.0-SNAPSHOT --> org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT
> ```
> 
> I'm not sure it's caused by this one here but looks suspiciously connected.

This looks lile caused by this PR.

Should we quickly revert it before next build?

akurtakov commented:
> > I just saw aggregator build failing https://ci.eclipse.org/platform/job/eclipse.platform.releng.aggregator/job/PR-1639/1/console :
> > ```
> > 22:23:57  org.apache.maven.ProjectCycleException: The projects in the reactor contain a cyclic reference: Edge between 'Vertex{label='eclipse.platform.repository:eclipse.platform.repository:4.31.0-SNAPSHOT'}' and 'Vertex{label='org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT'}' introduces to cycle in the graph org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT --> org.eclipse.pde:org.eclipse.pde.build.tests:1.4.300-SNAPSHOT --> eclipse.platform.repository:eclipse.platform.repository:4.31.0-SNAPSHOT --> org.eclipse.sdk.feature:org.eclipse.sdk.tests:4.31.0-SNAPSHOT
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm not sure it's caused by this one here but looks suspiciously connected.
> 
> This looks lile caused by this PR.
> 
> Should we quickly revert it before next build?

I'm 90% that this will break the build.

iloveeclipse commented:
So let revert it.

iloveeclipse commented:
> So let revert it.

I see no possibility to do this from the github UI?

If I see it right, we only need revert for https://github.com/eclipse-pde/eclipse.pde/commit/6e5138acab9998128905e533edf5d573b58196a3

iloveeclipse commented:
> > So let revert it.
> 
> I see no possibility to do this from the github UI?
> 
> If I see it right, we only need revert for [6e5138a](https://github.com/eclipse-pde/eclipse.pde/commit/6e5138acab9998128905e533edf5d573b58196a3)

=> https://github.com/eclipse-pde/eclipse.pde/pull/985

iloveeclipse commented:
@laeubi : it is unfortunate we've had to revert this PR, but it has shown a potential to enable PDE tests. 
If you have an idea how to make it without breaking aggregator build, it would be great.

laeubi commented:
I'll give it a try later on, basically I think one can use a profile that we only enable when verification builds run (like the `tck` profile we already have) so it is not executed in a full aggregator build.

HannesWell commented:
> I'll give it a try later on, basically I think one can use a profile that we only enable when verification builds run (like the `tck` profile we already have) so it is not executed in a full aggregator build.

Sounds good.
I also had a quick look back then and wondered why there is even a cycle?
`org.eclipse.pde.build.tests` is not a fragment to pde.build and should not be contained in any feature. Therefore I wonder why the sdk had a reference to it so that a cycle can be formed?

Another point is that it looked like the activated tests only use what have been build with the I-builds (see https://github.com/eclipse-pde/eclipse.pde/pull/984) and if I understand the config right, this is no surprise since the test-runtime is p2 provisioned?
